cassandra-driver/index.d.ts

```diff
@@ -506,11 +506,22 @@ export interface Client extends events.EventEmitter {
   metadata: metadata.Metadata;
 
   batch(queries: Array<string> | Array<{ query: string, params?: any }>, options: QueryOptions, callback: ResultCallback): void;
+  batch(queries: Array<string> | Array<{ query: string, params?: any }>, callback: ResultCallback): void;
+  batch(queries: Array<string> | Array<{ query: string, params?: any }>, options?: QueryOptions): Promise<types.ResultSet>;
   connect(callback: Callback): void;
+  connect(): Promise<void>;
   eachRow(query: string, params?: any, options?: QueryOptions, rowCallback?: Callback, callback?: Callback): void;
-  execute(query: string, params?: any, options?: QueryOptions, callback?: ResultCallback): void;
+  execute(query: string, params: any, options: QueryOptions, callback: ResultCallback): void;
+  execute(query: string, options: QueryOptions, callback: ResultCallback): void;
+  execute(query: string, params: any, callback: ResultCallback): void;
+  execute(query: string, callback: ResultCallback): void;
+  execute(query: string, params: any, options: QueryOptions): Promise<types.ResultSet>;
+  execute(query: string, options: QueryOptions): Promise<types.ResultSet>;
+  execute(query: string, params?: any): Promise<types.ResultSet>;
+  execute(query: string): Promise<types.ResultSet>;
```

Comment:
I think the different overloads expressed in TypeScript would be something like:
Callback-based
```javascript
execute(query: string, params: any, options: QueryOptions, callback: ResultCallback): void;
execute(query: string, params: any, callback: ResultCallback): void;
execute(query: string, callback: ResultCallback): void;
```
Promise based
```javascript
execute(query: string, params?: any, options?: QueryOptions): Promise<types.ResultSet>;
```

It's not allowed to specify the query options without specifying the parameters.